scylla-cdc/src/log_reader_builder.rs

```diff
@@ -0,0 +1,310 @@
+use std::sync::Arc;
+
+use anyhow::anyhow;
+use std::cmp::max;
+use std::time::Duration;
+use std::time::SystemTime;
+
+use futures::future::RemoteHandle;
+use futures::stream::{FusedStream, FuturesUnordered, StreamExt};
+use futures::FutureExt;
+use scylla::Session;
+
+use crate::cdc_types::GenerationTimestamp;
+use crate::consumer::ConsumerFactory;
+use crate::log_reader::CDCLogReader;
+use crate::stream_generations::GenerationFetcher;
+use crate::stream_reader::StreamReader;
+
+const SECOND_IN_MILLIS: i64 = 1_000;
+const SLEEP_INTERVAL: i64 = SECOND_IN_MILLIS / 10;
+const WINDOW_SIZE: i64 = SECOND_IN_MILLIS / 10 * 3;
+const SAFETY_INTERVAL: i64 = SECOND_IN_MILLIS / 10;
+
+enum NewCDCLogReaderError {
+    TableName,
+    KeyspaceName,
+    Session,
+    ConsumerFactory,
+}
+
+impl NewCDCLogReaderError {
+    fn as_str(&self) -> &'static str {
```

Comment:
It's more idiomatic to implement the [`Display`](https://doc.rust-lang.org/stable/std/fmt/trait.Display.html) trait and add `#[derive(Debug)]` for errors rather than introducing custom methods like that.